### ISSUE TITLE
Investigate why Java 11 builds (Azure Dev Pipelines) take 30-50% more time than Java 8

### ISSUE THREAD

**Issue by planetf1** [odpi/egeria#1818]
If one looks at recent builds such as:
https://dev.azure.com/odpi/Egeria/_build/results?buildId=1095&view=results

it's noted that the Java 8 build is consistently faster - noticeably so. Between 30-50%, and typically at least 5 mins more. This slows down merging of PRs when processing a backlog

Need to investigate cc: @bramwelt

**Comment by planetf1** [comment]
A few observations
 a) We are still creating the 'large' assembly artifact containing all egeria code which is no longer needed given our change to using finer grained artifacts from artifactory instead. - we could likely save 45s+ on both builds by removing this 
 b) On that particular build the repo upload took several minutes on java 11 side, less than 1 on 8 - but this is probably rather variable given upload times
 c) Whilst the overall maven elapsed time is indeed where we see a difference, the individual modules don't seem to differ that much
 d) The config for the jobs seems the same from the pipeline definition, differing only in java version. The same task version is used. The java compiler used is from azul systems but is at current levels (8/222, 11.0.4). These builds are openjdk based

**Comment by planetf1** [comment]
A few local tests (MBP 2016)
 - OpenJDK/hotspot 11.0.5 : 13 minutes -- and this is repeated
 - OpenJDK/hostspot 1.8.0.222 : 8 minutes
 - OpenJDK/openJ9 11.0.5 : 8 minutes

I'd like to do many more iterations to be sure - and this test was on MacOS. The build environment uses ubuntu, with a openjdk based java (Azul - default in Azure)

**Comment by planetf1** [comment]
Suggestion from openjdk team - 'Default GC implementation changes to G1 GC. ' wrt jdk 11
-> will try using same GC as 8

**Comment by planetf1** [comment]
Changing GC had no noticeable difference.

Some research including https://medium.com/@john_freeman/faster-maven-builds-with-maven-opts-822cdc82fa85 suggests a few other options.

"-XX:TieredStopAtLevel=1" has had a noticeable affect, but still not as fast as 8 (13-> 11 vs 8 minutes). There is significant use of this option by various projects (google search), and the impact is to limit hit optimization, which can be valuable for short running tasks (like each compile).

Proposal is to add this to the build environment to speed up PR validation.

Keeping this open to identify other improvements

**Comment by bramwelt** [comment]
Currently the builds use Java implementations supplied by the [VM](https://github.com/microsoft/azure-pipelines-image-generation/blob/master/images/linux/Ubuntu1804-README.md) (ubuntu-18.04 Azul Zulu OpenJDK).
Another suggestion would be to see if building within containers would show any improvement. That would also allow you to target a specific (or multiple) Java implementation(s).

**Comment by planetf1** [comment]
Azul is openjdk based too - it was interesting to see that openjdkz-hotspot builds locally are similarly affected. The suggestion above seems harmless, though I don't know how significant the improvement might be, Worth trying? Any concerns?

Any thoughts on the containers? Not checked docs to see how much work. I guess we won't know without trying ...

**Comment by planetf1** [comment]
From the 1 build that's run since the above commit, there was no difference to timings.
The option looks correct according to https://docs.microsoft.com/en-us/azure/devops/pipelines/tasks/build/maven?view=azure-devops

Leaving open as this remains a concern.

**Comment by planetf1** [comment]
it's possible the zing vm may not support this option (but may have others) - see https://docs.azul.com/zing/UseZVM_CommandLineOptionsZVM.htm .

However `./java -XX:+PrintFlagsFinal` does show this option (default 4 as expected)

**Comment by planetf1** [comment]
Note that the gradle build is parallel, and a lot quicker.
There are various threads with people reporting java 11 is slower for compile, but no strong conclusions. It does seem consistent.
At this point I can't see we can actually do anything about it -- a few tweaks to VM settings for the compile made at best only small changes, and really need more testing to ensure no behaviour change and even to see if they are stastically significant.
The big improvement is to work in parallel, and so I am closing this issue
